Cuda fixes for tests

agibsonccc opened the pull request:
Decompress cuda ndarray when it exists
Misc samediff fixes with execution
Add missing cuda blas delegator resource class (basically no op) Fix null pointer in atomic allocator with empty arrays Disable paragraph vectors tests on cuda (takes too long and isn't really intended to be used with cuda anyways) Add back missing declarations for migration from old optimizers

## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [ X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

agibsonccc committed:
Cuda fixes for tests
Decompress cuda ndarray when it exists
Misc samediff fixes with execution
Add missing cuda blas delegator resource class (basically no op)
Fix null pointer in atomic allocator with empty arrays
Disable paragraph vectors tests on cuda (takes too long and isn't really intended to be used with cuda anyways)
Add back missing declarations for migration from old optimizers

agibsonccc committed:
Cuda fixes for tests
Decompress cuda ndarray when it exists
Misc samediff fixes with execution
Add missing cuda blas delegator resource class (basically no op)
Fix null pointer in atomic allocator with empty arrays
Disable paragraph vectors tests on cuda (takes too long and isn't really intended to be used with cuda anyways)
Add back missing declarations for migration from old optimizers
Add allocation tracker fixes
Fix compression allocation
Add workspace helpers for memory allocation with cuda

treo commented on nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/workspace/WorkspaceUtils.java:
```diff
@@ -152,7 +153,41 @@ public static int getShapeBufferRequireMemoryForWorkspace(INDArray arr) {
+     */
     public static int getTotalRequiredMemoryForWorkspace(INDArray arr) {
-        return getAligned(arr.length() * arr.dataType().width()) + getAligned(arr.shapeInfoJava().length * DataType.INT64.width());
+        if(!Nd4j.getBackend().getNDArrayClass().getName().toLowerCase().contains("cu")) {
```
This looks more like a quick hack that can break in unexpected ways.

Also you are missing a `g` in the documentation:
```
     * GPUS will only be:
     * etAligned(arr.length() * arr.dataType().width())
```

agibsonccc replied:
Fixed! I added new metric recording and then use those in the unit tests. There's some hacks that can't be avoided due to the "answer" being different depending on the backend but it's at least checked against the expected metrics.

treo commented on platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/linalg/workspace/DebugModeTests.java:
```diff
@@ -134,6 +133,7 @@ public void testSpillMode_2(Nd4jBackend backend) {
             assertEquals(0, ws.getDeviceOffset());
 
             // array buffer should be spilled now
+            int shapeBufferAdd = backend.getClass().getName().toLowerCase().contains("cu") ? 0 : WorkspaceUtils.getShapeBufferRequireMemoryForWorkspace(array);
```
Same note about the hack applies here too. 

If you need it in multiple places, put it into a helper method.

agibsonccc replied:
Same answer as above.

agibsonccc committed:
Cuda fixes for tests
Decompress cuda ndarray when it exists
Misc samediff fixes with execution
Add missing cuda blas delegator resource class (basically no op)
Fix null pointer in atomic allocator with empty arrays
Disable paragraph vectors tests on cuda (takes too long and isn't really intended to be used with cuda anyways)
Add back missing declarations for migration from old optimizers
Add allocation tracker fixes
Fix compression allocation
Add workspace helpers for memory allocation with cuda
Migrate nd4j-native-api memory manager classes to nd4j-api to integrate with unified profiler
Add more unified profiler metrics
Add more device information to unifiedprofiler

agibsonccc committed:
Cuda fixes for tests
Decompress cuda ndarray when it exists
Misc samediff fixes with execution
Add missing cuda blas delegator resource class (basically no op)
Fix null pointer in atomic allocator with empty arrays
Disable paragraph vectors tests on cuda (takes too long and isn't really intended to be used with cuda anyways)
Add back missing declarations for migration from old optimizers
Add allocation tracker fixes
Fix compression allocation
Add workspace helpers for memory allocation with cuda
Migrate nd4j-native-api memory manager classes to nd4j-api to integrate with unified profiler
Add more unified profiler metrics
Add more device information to unifiedprofiler
Fix allocation counts

agibsonccc committed:
Cuda fixes for tests
Decompress cuda ndarray when it exists
Misc samediff fixes with execution
Add missing cuda blas delegator resource class (basically no op)
Fix null pointer in atomic allocator with empty arrays
Disable paragraph vectors tests on cuda (takes too long and isn't really intended to be used with cuda anyways)
Add back missing declarations for migration from old optimizers
Add allocation tracker fixes
Fix compression allocation
Add workspace helpers for memory allocation with cuda
Migrate nd4j-native-api memory manager classes to nd4j-api to integrate with unified profiler
Add more unified profiler metrics
Add more device information to unifiedprofiler
Fix allocation counts
More test fixes

agibsonccc committed:
Cuda fixes for tests
Decompress cuda ndarray when it exists
Misc samediff fixes with execution
Add missing cuda blas delegator resource class (basically no op)
Fix null pointer in atomic allocator with empty arrays
Disable paragraph vectors tests on cuda (takes too long and isn't really intended to be used with cuda anyways)
Add back missing declarations for migration from old optimizers
Add allocation tracker fixes
Fix compression allocation
Add workspace helpers for memory allocation with cuda
Migrate nd4j-native-api memory manager classes to nd4j-api to integrate with unified profiler
Add more unified profiler metrics
Add more device information to unifiedprofiler
Fix allocation counts
More test fixes
Add external allocations
Remove unused cuda divider method
